govcd/nsxt_l2_vpn_tunnel_test.go

```diff
@@ -0,0 +1,179 @@
+// 1. It creates/gets/updates/deletes a SERVER type Tunnel, also the peer code (encoded configuration of the tunnel)
+// is saved for creation of CLIENT type tunnel.
+// 2. Creates/gets/updates/deletes a CLIENT type Tunnel
+func (vcd *TestVCD) Test_NsxtL2VpnTunnel(check *C) {
+	skipNoNsxtConfiguration(vcd, check)
+	skipOpenApiEndpointTest(vcd, check, types.OpenApiPathVersion1_0_0+types.OpenApiEndpointEdgeGatewayL2VpnTunnel)
+
+	org, err := vcd.client.GetOrgByName(vcd.config.VCD.Org)
+	check.Assert(err, IsNil)
+
+	nsxtVdc, err := org.GetVDCByName(vcd.config.VCD.Nsxt.Vdc, false)
+	check.Assert(err, IsNil)
+
+	edge, err := nsxtVdc.GetNsxtEdgeGatewayByName(vcd.config.VCD.Nsxt.EdgeGateway)
+	check.Assert(err, IsNil)
+
+	network, err := nsxtVdc.GetOrgVdcNetworkByName(vcd.config.VCD.Nsxt.RoutedNetwork, false)
+	check.Assert(err, IsNil)
+
+	// Get the auto-allocated IP of the Edge Gateway
+	localEndpointIp, err := edge.GetUsedIpAddresses(nil)
+	check.Assert(err, IsNil)
+
+	// SERVER Tunnel part
+	serverTunnelParams := &types.NsxtL2VpnTunnel{
+		Name:                    check.TestName(),
+		Description:             check.TestName(),
+		SessionMode:             "SERVER",
+		Enabled:                 true,
+		LocalEndpointIp:         localEndpointIp[0].IPAddress,
+		RemoteEndpointIp:        "1.1.1.1",
+		TunnelInterface:         "",
+		ConnectorInitiationMode: "ON_DEMAND",
+		PreSharedKey:            check.TestName(),
+		StretchedNetworks: []types.EdgeL2VpnStretchedNetwork{
+			{
+				NetworkRef: types.OpenApiReference{
+					Name: network.OrgVDCNetwork.Name,
+					ID:   network.OrgVDCNetwork.ID,
+				},
+			},
+		},
+		Logging: false,
+	}
+
+	serverTunnel, err := edge.CreateL2VpnTunnel(serverTunnelParams)
+	check.Assert(err, IsNil)
+	check.Assert(serverTunnel, NotNil)
+	AddToCleanupListOpenApi(serverTunnel.NsxtL2VpnTunnel.ID, check.TestName(),
+		fmt.Sprintf(types.OpenApiPathVersion1_0_0+
+			types.OpenApiEndpointEdgeGatewayL2VpnTunnel+
+			serverTunnel.NsxtL2VpnTunnel.ID, edge.EdgeGateway.ID))
+
+	// Save the peer code to create a Client tunnel for testing
+	peerCode := serverTunnel.NsxtL2VpnTunnel.PeerCode
+
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.Name, Equals, check.TestName())
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.Description, Equals, check.TestName())
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.SessionMode, Equals, "SERVER")
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.Enabled, Equals, true)
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.LocalEndpointIp, Equals, localEndpointIp[0].IPAddress)
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.RemoteEndpointIp, Equals, "1.1.1.1")
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.ConnectorInitiationMode, Equals, "ON_DEMAND")
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.PreSharedKey, Equals, check.TestName())
+
+	fetchedServerTunnel, err := edge.GetL2VpnTunnelById(serverTunnel.NsxtL2VpnTunnel.ID)
+	check.Assert(err, IsNil)
+	check.Assert(fetchedServerTunnel, DeepEquals, serverTunnel)
```

Comment:
So this means that `PeerCode` is always returned on GET, did I understood it right? (I thought it was only retrievable on creation)

Reply:
There is a quirky behavior which should be fixed in the next version of API, but basically to summarize everything right now:

The server tunnel asks for pre-shared key on creation and always returns both pre-shared key and peer code on GET requests.
The client tunnel asks for peer code on creation and *shouldn't* return any of those two fields, but it returns only the decoded pre-shared key